src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,822 @@
+            if (_exception != null && _context != null)
+            {
+                _context.CloseContext();
+            }
+
+            return _exception;
+        }
+
+        internal bool IsAuthenticated
+        {
+            get
+            {
+                return _context != null && HandshakeComplete && _exception == null && _remoteOk;
+            }
+        }
+
+        internal bool IsMutuallyAuthenticated
+        {
+            get
+            {
+                if (!IsAuthenticated)
+                {
+                    return false;
+                }
+
+                // Suppressing for NTLM since SSPI does not return correct value in the context flags.
+                if (_context.IsNTLM)
+                {
+                    return false;
+                }
+
+                return _context.IsMutualAuthFlag;
+            }
+        }
+
+        internal bool IsEncrypted
+        {
+            get
+            {
+                return IsAuthenticated && _context.IsConfidentialityFlag;
+            }
+        }
+
+        internal bool IsSigned
+        {
+            get
+            {
+                return IsAuthenticated && (_context.IsIntegrityFlag || _context.IsConfidentialityFlag);
+            }
+        }
+
+        internal bool IsServer
+        {
+            get
+            {
+                return _context != null && _context.IsServer;
+            }
+        }
+
+        internal bool CanGetSecureStream
+        {
+            get
+            {
+                return (_context.IsConfidentialityFlag || _context.IsIntegrityFlag);
+            }
+        }
+
+        internal TokenImpersonationLevel AllowedImpersonation
+        {
+            get
+            {
+                CheckThrow(true);
+                return PrivateImpersonationLevel;
+            }
+        }
+
+        private TokenImpersonationLevel PrivateImpersonationLevel
+        {
+            get
+            {
+                // We should suppress the delegate flag in NTLM case.
+                return (_context.IsDelegationFlag && _context.ProtocolName != NegotiationInfoClass.NTLM) ? TokenImpersonationLevel.Delegation
+                        : _context.IsIdentifyFlag ? TokenImpersonationLevel.Identification
+                        : TokenImpersonationLevel.Impersonation;
+            }
+        }
+
+        private bool HandshakeComplete
+        {
+            get
+            {
+                return _context.IsCompleted && _context.IsValidContext;
+            }
+        }
+
+        internal void CheckThrow(bool authSucessCheck)
+        {
+            if (_exception != null)
+            {
+                throw _exception;
+            }
+
+            if (authSucessCheck && !IsAuthenticated)
+            {
+                throw new InvalidOperationException(SR.net_auth_noauth);
+            }
+        }
+
+        //
+        // This is to not depend on GC&SafeHandle class if the context is not needed anymore.
```

Comment:
I'm not sure I understand the comment here.  Aside from the overall not quite getting it, I don't know if that ampersand is a typo (GCSafeHandle) or is being used as "and", but without spacing around it.

And, no matter what, I don't see any SafeHandles here.